pkg/controller/seed-controller-manager/constraint-controller/controller.go

```diff
@@ -0,0 +1,280 @@
+	return reconcile.Result{}, err
+}
+
+func constraintCreatorGetter(constraint *kubermaticv1.Constraint) reconciling.NamedKubermaticV1ConstraintCreatorGetter {
+	return func() (string, reconciling.KubermaticV1ConstraintCreator) {
+		return constraint.Name, func(c *kubermaticv1.Constraint) (*kubermaticv1.Constraint, error) {
+			c.Name = constraint.Name
+			c.Spec = constraint.Spec
+			return c, nil
+		}
+	}
+}
+
+func (r *reconciler) patchFinalizer(ctx context.Context, constraint *kubermaticv1.Constraint, action string) error {
+	oldconstraint := constraint.DeepCopy()
+
+	if action == "add" {
+		kuberneteshelper.AddFinalizer(constraint, finalizer)
+	} else if action == "remove" {
+		kuberneteshelper.RemoveFinalizer(constraint, finalizer)
+	}
+
+	if err := r.seedClient.Patch(ctx, constraint, ctrlruntimeclient.MergeFrom(oldconstraint)); err != nil {
+		return fmt.Errorf("failed to %s constraint finalizer %s: %v", action, constraint.Name, err)
+	}
+
+	return nil
+}
+
+func (r *reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, request reconcile.Request) error {
+
+	constraint := &kubermaticv1.Constraint{}
+
+	if err := r.seedClient.Get(ctx, request.NamespacedName, constraint); err != nil {
+		return ctrlruntimeclient.IgnoreNotFound(err)
+	}
+
+	if !constraint.DeletionTimestamp.IsZero() {
+		if !kuberneteshelper.HasFinalizer(constraint, finalizer) {
+			return nil
+		}
+
+		if err := r.cleanupConstraint(ctx, log, constraint); err != nil {
+			return err
+		}
+	}
+
+	if !kuberneteshelper.HasFinalizer(constraint, finalizer) {
+		if err := r.patchFinalizer(ctx, constraint, "add"); err != nil {
+			return err
+		}
+	}
+
+	constraintCreatorGetters := []reconciling.NamedKubermaticV1ConstraintCreatorGetter{
+		constraintCreatorGetter(constraint),
+	}
+
+	return r.syncAllClustersNS(ctx, log, constraint, func(seedClient ctrlruntimeclient.Client, constraint *kubermaticv1.Constraint, namespace string) error {
+		return reconciling.ReconcileKubermaticV1Constraints(ctx, constraintCreatorGetters, namespace, seedClient)
+	})
+}
+
+func (r *reconciler) cleanupConstraint(ctx context.Context, log *zap.SugaredLogger, constraint *kubermaticv1.Constraint) error {
+
+	if err := r.syncAllClustersNS(ctx, log, constraint, func(seedClient ctrlruntimeclient.Client, constraint *kubermaticv1.Constraint, namespace string) error {
+
+		log := log.With("constraint", constraint)
+		log.Debugw("cleanup processing:", namespace)
+
+		err := seedClient.Delete(ctx, &kubermaticv1.Constraint{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      constraint.Name,
+				Namespace: namespace,
+			},
+		})
+
+		return ctrlruntimeclient.IgnoreNotFound(err)
+
+	}); err != nil {
+		return err
+	}
+
+	if err := r.patchFinalizer(ctx, constraint, "remove"); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func isOPAEnabled(userCluster *kubermaticv1.Cluster) bool {
+	if userCluster.Spec.OPAIntegration != nil && userCluster.Spec.OPAIntegration.Enabled {
+		return true
+	}
+	return false
```

Comment:
You can simplify this by changing to :

```suggestion
	return userCluster.Spec.OPAIntegration != nil && userCluster.Spec.OPAIntegration.Enabled 
```